state_test.go

```diff
@@ -24,68 +71,307 @@ func TestState_FaultyNodes(t *testing.T) {
 		{7, 2},
 		{8, 2},
 		{9, 2},
+		{10, 3},
 		{99, 32},
 		{100, 33},
 	}
 	for _, c := range cases {
-		acct := []string{}
-		for i := 0; i < int(c.Network); i++ {
-			acct = append(acct, fmt.Sprintf("acct_%d", i))
-		}
+		validatorNodeIds := generateValidatorNodes(c.TotalNodesCount, "validator")
 		s := newState()
-		s.validators = newMockValidatorSet(acct)
+		s.validators = convertToMockValidatorSet(validatorNodeIds)
 
-		assert.Equal(t, s.MaxFaultyNodes(), int(c.Faulty))
+		assert.Equal(t, s.MaxFaultyNodes(), c.FaultyNodesCount)
+	}
+}
+
+func TestState_ValidNodesCount(t *testing.T) {
+	cases := []struct {
+		TotalNodesCount, ValidNodesCount int
+	}{
+		{1, 0},
+		{2, 0},
+		{3, 0},
+		{4, 2},
+		{5, 2},
+		{6, 2},
+		{7, 4},
+		{8, 4},
+		{9, 4},
+		{10, 6},
+		{100, 66},
+	}
+	for _, c := range cases {
+		validatorNodeIds := generateValidatorNodes(c.TotalNodesCount, "validator")
```

Comment:
Do you ever need the result of this function `generateValidatorNodes`? It seems just an extra useless step.

Reply:
Result of function is added just because of readability.

Do you have this change in mind
convertToMockValidatorSet(generateValidatorNodes(c.TotalNodesCount, "validator"))
or there is more elegant way to rewrite that part? TBH, not sure whether we can avoid having generateValidatorNodes function, although I agree that it is pretty dummy.

Reply:
Second approach would be to instantiate `testerAccountPool` and to delete `generateValidatorNodes`. Although it could be overkill, could be more elegant way in terms of code brevity. :slightly_smiling_face: